Fix issue with improper notifications in propagate function

vaeleborne opened the pull request:
Fixes: #83160 

The issue was two-fold. First: NOTIFICATION_PAUSED and NOTIFICATION_UNPAUSED were not being sent if Process Mode was set to Always, see #104778, secondly and directly related in the logic, the wrong notification was being sent when Process Mode was set to when paused.

For Clarity: Now when Process Mode is Always, When_Paused, or Pausable, if when we pause the correct value of NOTIFICATION_PAUSED (14) is sent, when unpausing NOTIFICATION_UNPAUSED (15) is sent.

This PR corrects the logic and ensures that the proper notification is being sent, and is sent in all process modes it should be, (so not in disabled). 

Tested by running the mrp from #104778 

EDIT: Per comments, reworked so notifications are only being sent when Process Mode is either When_Paused or Pausable and ensured that they are both being sent the correct notification each time now by testing the p_enable to determine which message to send rather than checking the prev_can_process and next_can_process variables.

vaeleborne committed:
Fixed issue with improper notifications of _propagate_pause_notification(bool p_enable)

vaeleborne committed:
Fixed typo

AThousandShips commented:
This is not a bug, it's intended behavior IMO, the notification isn't about the tree, but about the node, so it should _not_ be sent when set to always, see the documentation:
> Notification received when the node is paused. See process_mode.

Now the part about the paused mode _is_ potentially a bug

vaeleborne commented:
> This is not a bug, it's intended behavior IMO, the notification isn't about the tree, but about the node, so it should _not_ be sent when set to always, see the documentation:
> 
> > Notification received when the node is paused. See process_mode.
> 
> 
> 
> Changing this breaks compatibility as the current behavior is the expected one
> 
> 
> 
> Now the part about the paused mode _is_ potentially a bug

That makes sense, I could take another look then without referencing the tree directly to see about that potential bug and correct the code style to conform.

vaeleborne committed:
Refactored to only target incorrect notification being sent

Removed check for process mode always as not sending a signal is expected behavior. Corrected logic so that the correct signal is sent in both process_mode_pausable and when_paused.
